---
title: "SCclust T10 Tutorial"
author: "Alex Krasnitz, Jude Kendall, Junyan Song, Lubomir Chorbadjiev"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    toc: yes
  pdf_document:
    number_sections: yes
    toc: yes
header-includes:
- \usepackage[colorinlistoftodos]{todonotes}
- \hypersetup{colorlinks=true,linkcolor=blue!30!black}
- \geometry{rmargin=1.5in}
vignette: |
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

The *SCclust* package implements feature selection based on
breakpoints, permutations for FDRs for Fisher test p-values and identification
of the clone structure in single cell copy number profiles.

In this tutorial we show how to use *SCclust* package using data, prepared by
*sgains* pipeline as described in 
[Example usage of sGAINS pipeline](https://github.com/KrasnitzLab/sgains/blob/master/docs/tutorial-navin2011.md). 
*SCclust* package is called as the last step in processing data from *sgains* pipeline. In this tutoral
we show how *SCclust* package could be used independently from *sgains* pipeline.

We assume that you have an R environment and have installed *SCclust* package as described in the `README.md`.

# Data

## Data for the T10 case

This tutorial is based on data published in:
[Navin N, Kendall J, Troge J, et al. Tumor Evolution Inferred by Single 
Cell Sequencing. 
Nature. 2011;472(7341):90-94. doi:10.1038/nature09807.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4504184/)
In particular we will use the data for polygenomic breast tumor T10 case available from SRA.
Description of samples for T10 could be found in
[Supplementary Table 1 | Summary of 100 Single Cells in the Polygenomic Tumor 
T10](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4504184/bin/NIHMS706545-supplement-Supplement_Table_1.xls)

We are going to run *SCclust* package on prepared by *sgains* pipeline [`varbin` step](https://github.com/KrasnitzLab/sgains/blob/master/docs/tutorial-navin2011.md#varbin-step).
You can go through all the step in [*sgains* T10 tutorial](https://github.com/KrasnitzLab/sgains/blob/master/docs/tutorial-navin2011.md)
and prepare this data.

For the purposes of this tutorial we recomend you to download already prepared `varbin` data from 
[example data](https://github.com/KrasnitzLab/SCclust/releases/download/v1.0.0RC3/navin_t10_varbin_data.tar.gz).
Apart from `varbin` T10 data you will need the binning scheme used in the analysis, that could be found [here](https://github.com/KrasnitzLab/SCclust/releases/download/v1.0.0RC3/hg19_R50_B20k_bins_boundaries.txt.gz).
And also we will need `cytoBand.txt` for HG19 that you can download it from UCSC Genome Browser.

## Collect the Neccessary Data

Let us create a directory, where to store all the data used in this tutorial:

```bash
mkdir T10data
cd T10data
```

and let us download and extract T10 `varbin` data:

```bash
wget -c \
  https://github.com/KrasnitzLab/SCclust/releases/download/v1.0.0RC3/navin_t10_varbin_data.tar.gz
tar zxvf navin_t10_varbin_data.tar.gz
rm navin_t10_varbin_data.tar.gz
```

Let us also download and extract the binning scheme used in preparation of `varbin` data:

```bash
wget -c \
  https://github.com/KrasnitzLab/SCclust/releases/download/v1.0.0RC3/hg19_R50_B20k_bins_boundaries.txt.gz
gunzip hg19_R50_B20k_bins_boundaries.txt.gz
```

And finally let us download the `cytoBand.txt` for Human reference genome *hg19*:

```bash
wget -c \
  http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/cytoBand.txt.gz
gunzip cytoBand.txt.gz
```

Our data directory should have following structure:

```bash
.
|-- T10data
    |-- cytoBand.txt
    |-- hg19_R50_B20k_bins_boundaries.txt
    |-- varbin
        |-- SRR052047.varbin.20k.txt
        |-- SRR052148.varbin.20k.txt
        |-- SRR053437.varbin.20k.txt
        ...
```

## Explore the Dowloaded Data

We are going to use *SCclust* package so let us load it:

```{r}
library("SCclust")
```

### Binning schema

\todo{Describe the data.}

```{r, results = "asis"}
gc_df <- read.csv("T10data/hg19_R50_B20k_bins_boundaries.txt", header = T, sep='\t')
knitr::kable(head(gc_df))
```

### Cytobands and Centromeres for HG19

\todo{Describe the data.}

```{r, results = "asis"}
cytobands <- read.csv("T10data/cytoBand.txt", header = F, sep='\t')
knitr::kable(head(cytobands))
```

The main reason we need `cytoBand.txt` is to get the location of centromeres. Since centromere areas contain a lot
of repetitive sequencies they are excluded from analysis when segmenting and clustering samples.

To find regions where centromeres are located we are using `calc_centroareas` function:

```{r, results = "asis"}
centroareas <- calc_centroareas(cytobands)
knitr::kable(head(centroareas, 5))
```


So, in `centroareas` for each chromosome we have the region where the centromere is located.

### Varbin Samples Data

\todo{Describe the data.}

For each `varbin` sample

```{r, results = "asis"}
sample_df <- read.csv("T10data/varbin/SRR052047.varbin.20k.txt", header=T, sep='\t')
knitr::kable(head(sample_df))
```

```{r, results = "asis"}
sample_df <- read.csv("T10data/varbin/SRR052148.varbin.20k.txt", header=T, sep='\t')
knitr::kable(head(sample_df))
```

# Auto Generated

Vignettes are long form documentation commonly included in packages. Because they are part of the distribution of the package, they need to be as compact as possible. The `html_vignette` output type provides a custom style sheet (and tweaks some options) to ensure that the resulting html is as small as possible. The `html_vignette` format:

- Never uses retina figures
- Has a smaller default figure size
- Uses a custom CSS stylesheet instead of the default Twitter Bootstrap style

## Vignette Info

Note the various macros within the `vignette` section of the metadata block above. These are required in order to instruct R how to build the vignette. Note that you should change the `title` field and the `\VignetteIndexEntry` to match the title of your vignette.

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
